### ISSUE TITLE
[oxc_semantic] Ecmascript TDZ - Temporal Dead Zone Violation

### ISSUE THREAD

**Issue by AraCoders** [oxc-project/oxc#12644]
given a code snippet like this:

```js
function outer() {
  let x = 1;

  function inner() {
    console.log(x + 2); // References outer 'x'
    let x = 3;          // Different 'x' binding
    return x;
  }

  return x;
}
```

oxc_semantic assumes that `x` reference in `console.log(x + 2);` is a reference for the this binding `let x = 3` not for `let x = 1;`. 

According to ES rules: let/const variables cannot be accessed before their declaration in the same scope. 

Node.js Runtime: Correctly throws ReferenceError: Cannot access 'x' before initialization
TypeScript/ESLint: Both detect this as an error

but it's totally valid code. and if you are processing minified js code (my use case), you will encounter many of these. 

the bug probably stems from: `resolve_references_for_current_scope()` at  `src/builder.rs` which resolves without checking if reference comes BEFORE declaration.

**Comment by overlookmotel** [comment]
The TDZ is pretty weird, so I do understand the confusion, but I'm afraid what you've said isn't correct.

In your example, `x` in `console.log(x + 2);` does *not* reference the outer `x`. It references the `x` binding created by `let x = 3` inside `inner`.

As you say, `let`/`const` variables cannot be accessed before their declaration. But that doesn't alter symbol resolution. The binding `x` in `inner` exists from the very start of the body of `inner`. But you just *can't access it* before the `let` declaration. This is the so-called "temporal dead zone" - "temporal" because it relates to the "time span" in which it's valid to access the binding. It's a *runtime* error, not a compile-time *syntax* error.

In your example above, there is no error, because `inner` is not called. But if you call `inner()`, you get the `ReferenceError: Cannot access 'x' before initialization` error you mentioned.

So `oxc_semantic` is correct here.

However... you're right that oxlint (at least with default rules) doesn't flag this problem. You said that ESLint does. What ESLint rule is it that catches this?
